ed_sensor_integration/tools/segmenter.cpp

```diff
@@ -184,6 +184,58 @@ void showSegmentationResults(const Snapshot& snapshot, const UpdateRequest& upda
 */
 // ----------------------------------------------------------------------------------------------------
 
+struct Quaternion {
+    double w, x, y, z;
+};
+
+struct EulerAngles {
+    double roll, pitch, yaw;
+};
+
+EulerAngles ToEulerAngles(Quaternion q) {
+    EulerAngles angles;
+
+    // roll (x-axis rotation)
+    double sinr_cosp = 2 * (q.w * q.x + q.y * q.z);
+    double cosr_cosp = 1 - 2 * (q.x * q.x + q.y * q.y);
+    angles.roll = std::atan2(sinr_cosp, cosr_cosp);
+
+    // pitch (y-axis rotation)
+    double sinp = 2 * (q.w * q.y - q.z * q.x);
+    if (std::abs(sinp) >= 1)
+        angles.pitch = std::copysign(M_PI / 2, sinp); // use 90 degrees if out of range
+    else
+        angles.pitch = std::asin(sinp);
+
+    // yaw (z-axis rotation)
+    double siny_cosp = 2 * (q.w * q.z + q.x * q.y);
+    double cosy_cosp = 1 - 2 * (q.y * q.y + q.z * q.z);
+    angles.yaw = std::atan2(siny_cosp, cosy_cosp);
+
+    return angles;
+}
+
+double getYaw(const geo::Mat3& rotation)
+{
+    // Get quaternion
+    geo::Quaternion geo_quaternion;
+    rotation.getRotation(geo_quaternion);
+
+    // Convert it to struct
+    Quaternion quaternion;
+    quaternion.x = geo_quaternion.getX();
+    quaternion.y = geo_quaternion.getY();
+    quaternion.z = geo_quaternion.getZ();
+    quaternion.w = geo_quaternion.getW();
+
+    // Get the Euler angles
+    EulerAngles angles = ToEulerAngles(quaternion);
+    ROS_DEBUG_STREAM("Matrix: " << rotation << " --> yaw: " << angles.yaw);
+    return angles.yaw;
+}
```

Comment:
If this is correct code, move it to geolib2

Reply:
In the scope of your BEP this is not required but if you would like to learn how to do this let us know. I can also imagine that there may be other libraries that can do this.